bridge/pier/syncer.go

```diff
@@ -212,17 +224,41 @@ func (syncer *Syncer) startSubscription(ctx context.Context, subscription ethere
 func (syncer *Syncer) processHeader(newHeader *types.Header) {
 	syncer.Logger.Debug("New block detected", "blockNumber", newHeader.Number)
 
-	latestNumber := newHeader.Number
+	// adding into queue
+	syncer.headerQueue.PushBack(&LightHeader{
+		Number: newHeader.Number,
+		Time:   newHeader.Time,
+	})
+
+	// current time
+	currentTime := uint64(time.Now().UTC().Unix())
+	confirmationTime := uint64(helper.GetConfig().TxConfirmationTime.Seconds())
+
+	var start *big.Int
+	var end *big.Int
+
+	// check start and end header
+	for syncer.headerQueue.Len() > 0 {
+		e := syncer.headerQueue.Front() // First element
+		h := e.Value.(*LightHeader)
+		if h.Time+confirmationTime > currentTime {
+			break
+		}
+
+		if start == nil {
+			start = h.Number
+		}
+		end = h.Number
 
-	// confirmation
-	confirmationBlocks := big.NewInt(0).SetUint64(helper.GetConfig().ConfirmationBlocks)
-	confirmationBlocks = confirmationBlocks.Add(confirmationBlocks, big.NewInt(1))
-	if latestNumber.Uint64() > confirmationBlocks.Uint64() {
-		latestNumber = latestNumber.Sub(latestNumber, confirmationBlocks)
+		syncer.headerQueue.Remove(e) // Dequeue
```

Comment:
Might be a better idea to Dequeue only when we are done processing these batch of blocks.
Say an error occurred during next execution steps, the block won't be re-tried, because it was already dequeued.

Reply:
In new bridge design will have retry mechanism. So, for now, I will keep it for cs-2004.